reqresp/requester.go

```diff
@@ -0,0 +1,218 @@
+	requesterBase
+	taskQueue string
+	worker    activityWorker
+}
+
+// Visible for testing
+type activityWorker interface {
+	RegisterActivity(interface{})
+	Start() error
+	Stop()
+}
+
+// Visible for testing
+var newWorker = func(c client.Client, taskQueue string) activityWorker {
+	return worker.New(c, taskQueue, worker.Options{})
+}
+
+func newRequesterActivityResponses(opts RequesterOptions) (*requesterActivityResponses, error) {
+	var r requesterActivityResponses
+	r.RequesterOptions = opts
+	r.pendingRequests = map[string]chan<- *Response{}
+	r.taskQueue = "requester-" + uuid.New()
+	r.worker = newWorker(opts.Client, r.taskQueue)
+	// Register the activity and start
+	r.worker.RegisterActivity(r.responseActivity)
+	if err := r.worker.Start(); err != nil {
+		return nil, fmt.Errorf("failed starting worker: %w", err)
+	}
+	return &r, nil
+}
+
+func (r *requesterActivityResponses) RequestUppercase(ctx context.Context, str string) (string, error) {
+	// Send request to use the response activity
+	return r.requestUppercase(ctx,
+		&Request{ID: uuid.New(), Input: str, ResponseActivity: "responseActivity", ResponseTaskQueue: r.taskQueue})
+}
+
+func (r *requesterActivityResponses) Close() { r.worker.Stop() }
+
+func (r *requesterActivityResponses) responseActivity(ctx context.Context, resp *Response) error {
+	// Get the channel to respond to
+	r.pendingRequestsLock.RLock()
+	respCh := r.pendingRequests[resp.ID]
+	r.pendingRequestsLock.RUnlock()
+	// We choose not to log or error if a response is not pending because it is
+	// normal behavior for a requester to have closed the context and stop waiting
+	if respCh == nil {
+		return nil
+	}
+
+	// Send non-blocking since the channel should have enough room. Technically
+	// during a situation where this worker was too busy for this activity to
+	// return, the responseActivity could be called again for the same response
+	// during retry from the other side. This will just result in a no-op since
+	// the channel does not have room.
+	select {
+	case respCh <- resp:
+	default:
+	}
+	return nil
+}
+
+// Visible for testing
+var tickerFreq = 3 * time.Second
+
+type requesterQueryResponses struct {
+	requesterBase
+	ticker     *time.Ticker
+	pollCtx    context.Context
+	pollCancel context.CancelFunc
+}
+
+func newRequesterQueryResponses(opts RequesterOptions) *requesterQueryResponses {
+	var r requesterQueryResponses
+	r.RequesterOptions = opts
+	r.pendingRequests = map[string]chan<- *Response{}
+	// Query every so often
+	r.ticker = time.NewTicker(tickerFreq)
+	r.pollCtx, r.pollCancel = context.WithCancel(context.Background())
+	// Start poller
+	go r.poll()
+	return &r
+}
+
+func (r *requesterQueryResponses) RequestUppercase(ctx context.Context, str string) (string, error) {
+	// Send request
+	return r.requestUppercase(ctx, &Request{ID: uuid.New(), Input: str})
+}
+
+func (r *requesterQueryResponses) Close() {
+	r.ticker.Stop()
+	r.pollCancel()
+}
+
+func (r *requesterQueryResponses) poll() {
```

Comment:
Why do it here instead of just after the Signal call? That would be much more simple.

Reply:
So that the requester can be used for many calls. Consider having one "requester" for an entire microservice that handles many requests. By polling in this way, you can get multiple responses as a time. Note how the query accepts a collection of IDs we are looking for responses to.

While this sample's CLI is ephemeral, the requester is built for long-term thread-safe use.

I could also call poll immediately after signal and handle debouncing if many happened at a time, but this was the simplest logic while keeping the response multiplexing functionality. If it's worth it, I can send the query right away per request.